### PR TITLE
Cisco support in Ansible

### PR DESCRIPTION
**Pull request by bennojoy**
## Configuring Cisco Devices with Ansible

This PR introduces a cisco transport plugin and a module which allows users/network_admins to configure
Cisco devices via Ansible, The implementation logic is quite straight forward, Network users/admins 
can create template files for their usual network configurations and instruct the module to execute
those configurations in all or certain subset of thier network infrastructure. 
The template file is parsed by jinja and any variable substituations are carried out. The module has been tested on
Cisco IOS as well as NX-OS.   

The Cisco transport plugin supports telnet and ssh as means to communicating with the device, It also 
supports enabling privilege mode transition(enable) on the target devices. Transport type can be set
at individual device level, so if a device uses telnet instead of ssh while making the entry for the device in inventory
the user can add a variable similar to the follwing "<device name> cisco_tranport=telnet", and plugin would use
telnet as the transport.

Following is an example where a Cisco routers interface is configured.

```
Inventory:
     [ios]
     router1 cisco_transport=telnet ansible_ssh_host=192.168.2.20

playbook:
    - hosts: ios
      user: ben
      connection: cisco
      vars:
        interface_name: "f0/1"
        ip_address: "10.0.0.3"
        netmask: "255.255.255.0"
        gather_facts: no
      tasks:
        - name: Cisco command
          action: cisco src=interface.j2 enable=yes enable_pass="foobar" commit="startup-config"

Interface.j2 file:
    conf t
    int {{ interface_name }}
    ip address {{ ip_address }} {{ netmask }}
    exit
    exit


Execution:

    ansible-playbook -i hosts cisco.yml  -k -vv
```

More information about the module is available via the module documentation.
Note: Currently the module is non idempotent, but i am contemplating to make it idempotent by gathering facts from network devices before the module is run and those facts can used to conditionally execute the task "when:".

Feedbacks and suggestion are welcome.

### CODE REVIEW

**Comment by blocke** [comment]
A couple very minor comments after taking a quick glance:

pexpect is a dependency that isn't bundled with Python and there should probably be a more friendly error message when the import fails.  Also probably needs a note in the documentation.

As you stated the module is not idempotent and does not handle what "state" (non-enabled, enabled, config, if-config, etc) the connected session is currently in.  It might be better to call this cisco_raw and conceptually treat it similarly to the raw module.  Just an idea.

The runner and connection code is a great first step and the idea is exciting.  Making the state handling a bit more robust and adding a fact collector would be a great starting point for idempotent cisco modules.  Setting up a serial connection type would be great idea too.

I'll try to give it a whirl in my environment at work.

**Comment by bennojoy** [comment]
Thanks for the feedback, Will incorporate the import check and rename the module to cisco_raw as you suggested.
could you elaborate a bit on the "serial connection type" 
- Benno

**Comment by blocke** [comment]
Actually the more I think about it the more of a bad idea renaming it to cisco_raw would be.  If you're planning to add more to this overtime then having it as a subcommand of a "cisco" module would probably be a better idea.  I've seen this pattern in other modules such as the digital_ocean one.

I've been wishing for something like this for a while and you've demonstrated it's definitely possible which is exciting.  I played around with how I might personally want to structure a playbook managing a switch and came up with this:

https://gist.github.com/blocke/6812723

I think if you have an action= or command= param then you can easily just move your existing raw template functionality to something like:

```
cisco: action=raw template=foo.j2
```

And people can add other action with idempotent sub-actions within the monolithic cisco module.  That seems to be how many of the cloud provisioning modules are structured. 

As for my mention of a serial connection type I noticed that there doesn't seem to be anything to prevent reusing the bulk of your connector code to also target the provisioning of cisco devices over a directly connected serial console link.

I've played around with using the python pexpect module over a serial link and this is basically the code to make it happen:

```
import fdpexpect
import serial

...
# Initalize serial port
serialdev = serial.Serial(serial_file, baud, bits, parity, timeout=None)
serialfd = serialdev.fileno()

# Flush current contents of serial Input buffer
serialdev.flushInput()

# Pass file descriptor for open serial port to pexpect
child = fdpexpect.fdspawn(serialfd)

```

fdexpect is part of the pexpect package and works on open file descriptors instead of interacting with spawning child processes.  You initialize the serial link then pass the file descriptor for the serial link to fdpexpect.  From that point forward fdpexpect has the same API as pexpect. 

The only catches here are that you have to be mindful when waiting for prompts that things will be slower and when you open that serial connection you will have no idea what mode the prompt is currently in (are you enabled?  is there a password prompt?).  The prompt expect portion would have to be made more robust and made to figure out how to get to the desired prompt mode before executing commands.  Beyond that the cisco module would operate exactly the same between going over SSH and going over a local serial console link.

Thoughts?

**Comment by bennojoy** [comment]
yeah, i guess we will keep the name as cisco itself for now. for your idea of idempotency support by adding 'action' and 'type'  seems doable but it would require to parse the running config and then check it against the lines specified and it might get tricky, but worth giving a try.
For idempotency i was thinking in terms like we would gather facts from the running config and populate values like 
'ansible_nexus5k_int_0_0': 1.1.1.1 for the target devices
and in the task we would do something like:

cisco: src=interface.j2
when: ansible_nexus5k_int_0_0 != n5k_address.

I would check the options as soon i get some free cycles to work on it. meanwhile if you can give it a try that would be great too.

For the serial option, as you mentioned the connection plugin can easily be made to support serial connection, i may not have the bandwidth to try it out now, but feel free to send in a patch if you can, else i will incorporate that later.

**Comment by blocke** [comment]
Yeah.  Being idempotent will require quite a bit of work (including parsing of running-configs) but it fits in more with the zen of ansible more than just raw.  Raw definitely has it's place and time but in order to do two raws on the same connection the end user is going to have to keep track of prompt state and how many exits they need and such.  In my mind one of the reasons I chose ansible is to avoid having to manually manage the shell or be aware of expect magic occurring on the target system.

If the action/type style is implemented then you can probably get raw mode merged first then idempotent features can be added over time as separate pull requests that don't break existing playbooks.  I figure it would be more encouraging for people to contribute small pieces then wait for a large monolithic pull request.  But others may disagree with me on that.

As for the serial thing I'll see if I can't get it working over the next couple of days.  Thanks again for kicking this off!

**Comment by candlerb** [comment]
What about using rancid to provide the transport layer? Then it could be used to manage a whole plethora of different devices. People who are already pulling configs with rancid would be able to push changes with minimal extra work. It could read the existing rancid inventory too.

Of course, it would have to be called "ransible" :-)

File: lib/ansible/runner/action_plugins/cisco.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            raise errors.AnsibleError("Cisco module is only usable in playbooks")
+
+        # load up options
+        options  = {}
+        if complex_args:
+            options.update(complex_args)
+        options.update(utils.parse_kv(module_args))
+
+        source         = options.get('src', None)
+	transport      = inject.get('cisco_transport', 'ssh')
```
Comment (resmo) [inline lib/ansible/runner/action_plugins/cisco.py:44]: indention?

File: library/net_infrastructure/cisco [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# this is a virtual module that is entirely implemented server side
+
+DOCUMENTATION = '''
+---
+module: cisco 
+short_description: Templates a cisco device configuration file and passes it on to cisco transport for execution.
+description:
+    - This module along with the cisco tranport plugin allows users/network_admins to configure
+      cisco devices via ansible, The implementation logic is quite straight forward, Network users/admins 
+      can create template files for thier usual network configurations and instruct the module to execute
```
Comment (ghost) [inline library/net_infrastructure/cisco:10]: their

File: library/net_infrastructure/cisco [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      those configurations in all or certain subset of thier network infrastructure. The template file
+      is parsed by jinja and any variable substituations are carried out. This module has been tested on
+      Cisco IOS as well as NX-OS.   
+    - The cisco transport plugin supports telnet and ssh as means to communicating with the device, It also 
+      supports enabling preveilege mode transition(enable) on the target devices. Transport type can be set
```
Comment (borgified) [inline library/net_infrastructure/cisco:15]: privileged

File: library/net_infrastructure/cisco [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      at individual device level, so if a device uses telnet instead of ssh while making the entry for the device in inventory
+      add a variable similar to the follwing "<device name> cisco_tranport=telnet", and plugin would use
```
Comment (borgified) [inline library/net_infrastructure/cisco:17]: following

**Comment by mpdehaan** [comment]
We are not going to include this in this time, others are definitely free to use if they want -- though I would want to have an API for managing the device to be willing to support this into the future -- see similar examples with F5 and Arista -- though I'll admit this is definitely the problem with Cisco as is, I hear they are moving towards the right direction.

**Pull request closed without merging**